chatkit-core/src/main/kotlin/com/pusher/chatkit/users/api/UserSubscriptionDispatcher.kt

```diff
@@ -0,0 +1,56 @@
+package com.pusher.chatkit.users.api
+
+import com.pusher.chatkit.rooms.api.JoinedRoomApiTypeMapper
+import com.pusher.chatkit.rooms.state.JoinedRoomsStateDiffer
+import com.pusher.chatkit.state.JoinedRoom
+import com.pusher.chatkit.state.JoinedRoomsReceived
+import com.pusher.chatkit.state.LeftRoom
+import com.pusher.chatkit.state.RoomDeleted
+import com.pusher.chatkit.state.RoomUpdated
+import org.reduxkotlin.Dispatcher
+
+internal class UserSubscriptionDispatcher(
+    private val joinedRoomsStateDiffer: JoinedRoomsStateDiffer,
+    private val joinedRoomApiTypeMapper: JoinedRoomApiTypeMapper,
+    private val dispatcher: Dispatcher
+) {
+
+    internal fun onEvent(event: UserSubscriptionEvent) {
+        when (event) {
+            is UserSubscriptionEvent.InitialState -> {
+
+                if (joinedRoomsStateDiffer.stateExists()) {
+                    joinedRoomsStateDiffer.toActions(
+                        joinedRoomApiTypeMapper.toRoomInternalTypes(event.rooms),
+                        joinedRoomApiTypeMapper.toUnreadCounts(event.readStates)
+                    ).forEach {
+                        action -> dispatcher(action) }
+                } else {
+                    dispatcher(
+                        JoinedRoomsReceived(
+                            joinedRoomApiTypeMapper.toRoomInternalTypes(event.rooms),
+                            joinedRoomApiTypeMapper.toUnreadCounts(event.readStates)
+                        )
+                    )
+                }
+            }
+            is UserSubscriptionEvent.AddedToRoomEvent -> {
+                dispatcher(
+                    JoinedRoom(
+                        joinedRoomApiTypeMapper.toRoomInternalType(event.room),
+                        event.readState.unreadCount
+                    )
+                )
+            }
+            is UserSubscriptionEvent.RemovedFromRoomEvent -> {
+                dispatcher(LeftRoom(event.roomId))
+            }
+            is UserSubscriptionEvent.RoomUpdatedEvent -> {
+                dispatcher(RoomUpdated(joinedRoomApiTypeMapper.toRoomInternalType(event.room)))
```

Comment:
for joined rooms feature we should also handle `ReadStateUpdatedEvent`, extract unreadCount from it and dispatch `UnreadCountReceived` event which will be handled by its respective reducer with the change applied only if unreadCount differs

Reply:
actually it would be best to card it and have a separate PR, small atomic PRs that add value and are tested are nice :)

Reply:
:+1 have created the card to track this in leankit